core/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -545,6 +545,15 @@ public boolean ensureSynced(Location location) throws IOException {
                 ensureOpen();
                 return current.syncUpTo(location.translogLocation + location.size);
             }
+        } catch (AlreadyClosedException | IOException ex) {
+            if (current.getTragicException() != null) {
```

Comment:
we repeat this pattern in a couple of places and it's easy to get wrong. Maybe add a closeOnTragicEvent method?